pkg/skaffold/jib/jib.go

```diff
@@ -17,64 +17,163 @@ limitations under the License.
+	Inputs []string
+
+	// Ignore lists paths to files that should be ignored when checking for changes to rebuild
+	Ignore []string
+}
+
+// filesLists contains cached build/input dependencies
+type filesLists struct {
+	// FilesTemplate saves the most recent output of the Jib Skaffold files task/goal
+	FilesTemplate filesTemplate
+
+	// BuildFileTimes keeps track of the last modification time of each build file
+	BuildFileTimes map[string]time.Time
+}
+
+// watchedFiles maps from project name to watched files
+var watchedFiles = map[string]filesLists{}
+
+// getDependencies returns a list of files to watch for changes to rebuild
+func getDependencies(cmd *exec.Cmd, projectName string) ([]string, error) {
+	dependencyList := []string{}
+	template := watchedFiles[projectName].FilesTemplate
+	if len(template.Inputs) == 0 && len(template.Build) == 0 {
+		// Make sure build file modification time map is setup
+		if watchedFiles[projectName].BuildFileTimes == nil {
+			watched := watchedFiles[projectName]
+			watched.BuildFileTimes = make(map[string]time.Time)
+			watchedFiles[projectName] = watched
+		}
+
+		// Refresh dependency list if empty
+		if err := refreshDependencyList(cmd, projectName); err != nil {
+			return nil, err
+		}
+	} else {
+		// Walk build files to check for changes
+		if err := walkFiles(&template.Build, &template.Ignore, func(path string, info os.FileInfo) error {
+			if val, ok := watchedFiles[projectName].BuildFileTimes[path]; !ok || info.ModTime() != val {
+				return refreshDependencyList(cmd, projectName)
+			}
+			return nil
+		}); err != nil {
+			return nil, err
+		}
+	}
+
+	// Walk updated files to build dependency list
+	watched := watchedFiles[projectName]
+	if err := walkFiles(&watched.FilesTemplate.Inputs, &watched.FilesTemplate.Ignore, func(path string, info os.FileInfo) error {
+		dependencyList = append(dependencyList, path)
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+	if err := walkFiles(&watched.FilesTemplate.Build, &watched.FilesTemplate.Ignore, func(path string, info os.FileInfo) error {
+		dependencyList = append(dependencyList, path)
+		watched.BuildFileTimes[path] = info.ModTime()
+		return nil
+	}); err != nil {
 		return nil, err
 	}
 
-	cmdDirInfo, err := os.Stat(cmd.Dir)
+	sort.Strings(dependencyList)
+	return dependencyList, nil
+}
+
+// refreshDependencyList calls out to Jib to retrieve an up-to-date list of files/directories to watch
+func refreshDependencyList(cmd *exec.Cmd, projectName string) error {
+	stdout, err := util.RunCmdOut(cmd)
 	if err != nil {
-		return nil, err
+		return errors.Wrap(err, "failed to get Jib dependencies; it's possible you are using an old version of Jib (Skaffold requires Jib v1.0.2+)")
 	}
 
-	// Parses stdout for the dependencies, one per line
 	lines := util.NonEmptyLines(stdout)
+	for i := range lines {
+		if lines[i] == "BEGIN JIB JSON" {
+			// Found Jib JSON header, next line is the JSON
+			files := watchedFiles[projectName]
```

Comment:
1. I feel like this could move moved outside the loop alongside with the pointer to the `filesTemplate`, and you should be able to repeatedly unmarshal into the filestemplate (assuming the lines are disjunct...?) and then `watchedFiles[projectName] = file` will be unnecessary too on L122. 

2. A little explanation, or reference to the JIB file structure would be great too for better understanding.

Reply:
1. Not sure I follow. There's no need to repeatedly unmarshal, since all the JSON will always be on one line; the reason for the loop is to look for the "BEGIN JIB JSON" header, and once it's found, everything else will only execute once. And I think `watchedFiles[projectName] = files` would still be necessary to save the unmarshalled file lists; the reason for the
```
files := watchedFiles[projectName]
...&files.PathMap...
watchedFiles[projectName] = files
```
is because you can't get a pointer to a field inside a map, I guess?

2. Sure, I'll add this.